src/awkward/_connect/numba/layoutbuilder.py

```diff
@@ -0,0 +1,2547 @@
+        c.context, c.builder, value=val
+    )
+    data_obj = c.pyapi.from_native_value(typ.data, builder.data, c.env_manager)
+
+    out = c.pyapi.call_function_objargs(
+        from_buffer_obj,
+        (data_obj,),
+    )
+
+    # decref PyObjects
+    c.pyapi.decref(Numpy_obj)
+    c.pyapi.decref(from_buffer_obj)
+
+    c.pyapi.decref(data_obj)
+
+    return out
+
+
+def _from_buffer():
+    raise RuntimeError("_from_buffer Python function is only implemented in Numba")
+
+
+@numba.extending.type_callable(_from_buffer)
+def Numpy_from_buffer_typer(context):
+    def typer(buffer):
+        if isinstance(buffer, GrowableBufferType):
+            return NumpyType(buffer.dtype, parameters=None)
+
+    return typer
+
+
+@numba.extending.lower_builtin(_from_buffer, GrowableBufferType)
+def Numpy_from_buffer_impl(context, builder, sig, args):
+    out = numba.core.cgutils.create_struct_proxy(sig.return_type)(context, builder)
+    out.data = args[0]
+
+    if context.enable_nrt:
+        context.nrt.incref(builder, sig.args[0], args[0])
+
+    return out._getvalue()
+
+
+@numba.extending.overload(Numpy)
+def Numpy_ctor(dtype, parameters=None, initial=1024, resize=8.0):
+    if isinstance(dtype, numba.types.StringLiteral):
+        dt = np.dtype(dtype.literal_value)
+
+    elif isinstance(dtype, numba.types.DTypeSpec):
+        dt = numba.core.typing.npydecl.parse_dtype(dtype)
+
+    else:
+        return
+
+    def ctor_impl(dtype, parameters=None, initial=1024, resize=8.0):
+        panels = numba.typed.List([np.empty((initial,), dt)])
+        length_pos = np.zeros((2,), dtype=np.int64)
+        data = ak._connect.numba.growablebuffer._from_data(panels, length_pos, resize)
+
+        return _from_buffer(data)
+
+    return ctor_impl
+
+
+@numba.extending.overload_method(NumpyType, "_length_get", inline="always")
+def Numpy_length(builder):
+    def getter(builder):
+        return builder.data._length_pos[0]
+
+    return getter
+
+
+@numba.extending.overload_attribute(NumpyType, "dtype", inline="always")
+def Numpy_dtype(builder):
+    def get(builder):
+        return builder._data.dtype
+
+    return get
+
+
+@numba.extending.overload_attribute(NumpyType, "data", inline="always")
+def Numpy_buffer(builder):
+    def get(builder):
+        return builder._data
+
+    return get
+
+
+@numba.extending.overload_method(NumpyType, "append")
+def Numpy_append(builder, datum):
+    if isinstance(builder, NumpyType):
+
+        def append(builder, datum):
+            builder.data.append(datum)
+
+        return append
+
+
+@numba.extending.overload_method(NumpyType, "extend")
+def Numpy_extend(builder, data):
+    def extend(builder, data):
+        builder.data.extend(data)
+
+    return extend
+
+
+@numba.extending.overload_method(NumpyType, "snapshot")
+def Numpy_snapshot(builder):
+    def snapshot(builder):
+        # FIXME: returns 'numpy.ndarray'
+        return builder.data.snapshot()
+
+    return snapshot
+
+
```

Comment:
The `FIXME` says it: it's an incorrect implementation because the Python version returns some sort of an Awkward Array (`ak.Array` or `ak.contents.Content` subclass) and the Numba implementation returns a NumPy array. But you don't have to (and can't) implement these in the compiled context.

```suggestion
```

They also weren't consistently implemented, and only this one was tested, which is probably why it had to temporarily return a NumPy array (which is possible, instead of an Awkward Array, which is impossible).